python/cudf/cudf/core/udf/masked_typing.py

```diff
@@ -423,3 +429,203 @@ def generic(self, args, kws):
 
 for unary_op in unary_ops:
     cuda_decl_registry.register_global(unary_op)(MaskedScalarUnaryOp)
+
+
+# Strings functions and utilities
+def _is_valid_string_arg(ty):
```

Comment:
I'm a bit confused. It looks like the contents of strings_udf/_typing was moved into core/udf/strings_typing and then the old contents of strings_typing were moved into this file. Could you comment on the rationale for the current separation? Is there a distinction between strings and nullable strings logic being made?

Reply:
Yes! This is a great question. Our `MaskedType` extension is the type we use to carry around a value and a validity. When using `DataFrame.apply` or `Series.apply`, the scalar inside your UDF are really `MaskedType`:



```python
df = cudf.DataFrame(
    {
        'a':[1, 2, 3], 
        'b':['a','b','c']  # one string column 'b'
}) 
def f(row):
    x = row['a'] # MaskedType(int64)
    y = row['b'] # MaskedType(string_view)
    z = y.upper() # MaskedType(udf_string), not used just for demo
    return x + len(y)
```

I always thought the cleanest way of putting this together was to have the `string_view` and `udf_string` types actually implement the string methods, like `upper` or `len`, whereas the `MaskedType` that carries it around holds the nullability logic. At that point calling something like `len(MaskedType(string_view))` is programmed in the lowering to translate roughly to:

```python
len(MaskedType(string_view, valid=True)) = MaskedType(len(string_view), valid=True)
```
meaning it ends up taking the validity of the source. That would make our strings just another type of scalar that could be makes and everything would be great, ideally.

The requirement of the external package made a lot of this way harder though since we needed a lot of this to be optional. Before this PR we had this happening:

- string scalars and their ops defined in `strings_udf` 
- Masked string operations defined in `cudf` (in `strings_typing.py` and `strings_lowering.py` deliberately in a separate place so that it could be optionally imported/registered
- `cudf` optionally importing `strings_typing` and `strings_lowering` which contained all the overloads for _masked_ strings

Now what we have is:
- string scalars and their ops defined in `strings_typing.py` and `strings_lowering.py`
- Masked string operations defined in `masked_typing.py` and `masked_lowering.py` 
- Nothing is optional

I think the second way is a lot cleaner.

Reply:
Idle thought, not for implementation in this PR, is there a way we can take advantage of what looks to be the functorial nature of the `MaskedType` object to simplify the code and reuse that from `strings_typing`/`lowering` more systematically?

Reply:
Thanks for the great explanation! That makes sense, and the new way is certainly simpler without the old extra package.